kratos/processes/calculate_discontinuous_distance_to_skin_process.cpp

```diff
@@ -542,6 +659,216 @@ namespace Kratos
 		MathUtils<double>::CrossProduct(rIntObjNormal, rGeometry[1]-rGeometry[0], rGeometry[2]-rGeometry[0]);
 	}
 
+	template<std::size_t TDim>
+	void CalculateDiscontinuousDistanceToSkinProcess<TDim>::ComputeExtrapolatedEdgesIntersectionsIfIncised(
+        const Element& rElement,
+		unsigned int &rNumCutEdges,
+		array_1d<double, (TDim == 2) ? 3 : 6>& rCutEdgesRatioVector,
+		array_1d<double,3> &rExtraGeomNormal,
+		array_1d<double, (TDim == 2) ? 3 : 6>& rCutExtraEdgesRatioVector)
+	{
+		if (TDim == 2 && rNumCutEdges == 1) {
+			ComputeExtrapolatedGeometryIntersections(rElement, rNumCutEdges, rCutEdgesRatioVector, rExtraGeomNormal, rCutExtraEdgesRatioVector);
+		} else if (TDim == 3) {
+			if (rNumCutEdges == 3) {
+				// if all three cut edges share one node, then the element is intersected and not incised
+				if (CheckIfCutEdgesShareNode(rElement, rCutEdgesRatioVector)) {
+					return;
+				}
+				ComputeExtrapolatedGeometryIntersections(rElement, rNumCutEdges, rCutEdgesRatioVector, rExtraGeomNormal, rCutExtraEdgesRatioVector);
+			} else if (rNumCutEdges == 1 || rNumCutEdges == 2) {
+				ComputeExtrapolatedGeometryIntersections(rElement, rNumCutEdges, rCutEdgesRatioVector, rExtraGeomNormal, rCutExtraEdgesRatioVector);
+			}
+		}
+	}
+
+	template<std::size_t TDim>
+	void CalculateDiscontinuousDistanceToSkinProcess<TDim>::ComputeExtrapolatedGeometryIntersections(
+        const Element& rElement,
+		unsigned int& rNumCutEdges,
+		array_1d<double, (TDim == 2) ? 3 : 6>& rCutEdgesRatioVector,
+		array_1d<double,3>& rExtraGeomNormal,
+		array_1d<double, (TDim == 2) ? 3 : 6>& rCutExtraEdgesRatioVector)
+	{
+		const auto r_edges_container = rElement.GetGeometry().GenerateEdges();
```

Comment:
Not sure if possible but we should try to call the `GenerateEdges` only once as it affects the parallelism when creating the shared pointer to each edge geometry. Hence, if the usages are more than the one in here (which I think is the case) we should create the edges container once in the outer layer and pass it by reference.

Reply:
It is the case .. I'll try to do that.